ElectronClient/gui/NoteEditor/NoteBody/AceEditor/CodeMirror.tsx

```diff
@@ -0,0 +1,206 @@
+import * as React from 'react';
+import { useEffect, useImperativeHandle, useState,  useCallback, forwardRef } from 'react';
+
+const CodeMirror = require('codemirror');
+import 'codemirror/addon/comment/comment';
+import 'codemirror/addon/dialog/dialog';
+import 'codemirror/addon/edit/closebrackets';
+import 'codemirror/addon/edit/continuelist';
+import 'codemirror/addon/scroll/scrollpastend';
+import 'codemirror/addon/search/searchcursor';
+import 'codemirror/addon/search/search';
+
+import useListIdent from './utils/useListIdent';
+
+import 'codemirror/keymap/emacs';
+import 'codemirror/keymap/vim';
+
+import 'codemirror/mode/gfm/gfm';
+import 'codemirror/mode/xml/xml';
+// Modes for syntax highlighting inside of code blocks
+import 'codemirror/mode/python/python';
+
+export interface CancelledKeys {
+	mac: string[],
+	default: string[],
+}
+
+export interface EditorProps {
+	value: string,
+	mode: string,
+	style: any,
+	theme: any,
+	readOnly: boolean,
+	autoMatchBraces: boolean,
+	keyMap: string,
+	cancelledKeys: CancelledKeys,
+	onChange: any,
+	onScroll: any,
+	onEditorContextMenu: any,
+	onEditorPaste: any,
+}
```

Comment:
The editor should support the NoteBodyEditorProps (doesn't have to support all of them, but that should be the input).

Reply:
I've restructured the code a bit, so what you're looking at here is basically a react wrapper around codemirror. There is another component that handles the integration with Joplin. Confusingly I've renamed that other one to CodeMirror.tsx and called the wrapper Editor.tsx.